selfie.c

```diff
@@ -4053,12 +4198,741 @@ int main_emulator(int argc, int *argv, int *cstar_argv) {
+    
+    prev = list_get_entry_at(list, index - 1);
+    entry = list_entry_get_next(prev);
+    next = list_entry_get_next(entry);
+    list_entry_set_next(prev, next);
+    list_set_size(list, size - 1);
+
+    return entry;
+}
+
+int *list_get_entry_at(int *list, int index) {
+    int i;
+    int size;
+    int *head;
+    int *tail;
+ 
+    size = list_get_size(list);
+    // If it's the last element just return TAIL
+    if(index == size - 1) {
+        tail = list_get_tail(list);
+        return tail;
+    }
+
+    head = list_get_head(list);
+    i = 0;
+    while(i < size) {
+        if(i == index)
+            return head;
+        head = list_entry_get_next(head);
+        i = i + 1;
+    }
+
+    return 0;
+}
+
+int list_is_in_bounds(int *list, int index) {
+    int size;
+
+    size = list_get_size(list);
+    if(index < 0)
+        return 0;
+    else if(size <= index)
+        return 0;
+    return 1;
+}
+
+int list_is_empty(int *list) {
+    int size;
+
+    size = list_get_size(list);
+
+    return size == 0;
+}
+
+void list_swap(int *list, int index1, int index2) {
+    int *entry1;
+    int *entry2;
+    int *data1;
+    int *data2;
+
+    if(list_is_in_bounds(list, index1) == 0)
+        return;
+    if(list_is_in_bounds(list, index2) == 0)
+        return;
+
+    entry1 = list_get_entry_at(list, index1);
+    entry2 = list_get_entry_at(list, index2);
+    data1 = list_entry_get_data(entry1);
+    data2 = list_entry_get_data(entry2);
+
+    list_entry_set_data(entry1, data2);
+    list_entry_set_data(entry2, data1);
+}
+
+// Sorts list by the field at FIELD_NR
+void list_sort_by(int *list, int field_nr) {
+    int i;
+    int j;
+    int size;
+    int l;
+    int r;
+
+    size = list_get_size(list);
+    i = size;
+    while(i > 1) {
+        j = 0;
+        while(j < size - 1) {
+            l = *(list_entry_get_data(list_get_entry_at(list, j)) + field_nr);
+            r = *(list_entry_get_data(list_get_entry_at(list, j + 1)) + field_nr);
+            if(l > r)
+                list_swap(list, j, j + 1);
+            j = j + 1;
+        }
+        i = i - 1;
+    }
+}
+
+// Entry
+// Getters
+int *list_entry_get_next(int *entry) {
+    return (int*)*entry;
+}
+
+int *list_entry_get_prev(int *entry) {
+    return (int*)*(entry + 1);
+}
+
+int *list_entry_get_data(int *entry) {
+    return (int*)*(entry + 2);
+}
+
+// Setters
+void list_entry_set_next(int *entry, int *next) {
+    *entry = (int)next;
+}
+
+void list_entry_set_prev(int *entry, int *prev) {
+    *(entry + 1) = (int)prev;
+}
+
+void list_entry_set_data(int *entry, int *data) {
+    *(entry + 2) = (int)data;
+}
+
+// -----------------------------------------------------------------
+// ---------------------------- PROCESS ----------------------------
+// -----------------------------------------------------------------
+
+int *process_init(int id, int *registers, int *memory, int reg_hi, int reg_lo) {
+    int *process;
+
+    // process:
```

Comment:
see comment above concerning the data layout of the nodes. If you would add 2 words at the beginning of your process "structure" (prev, next), you could build lists inside that structure.